Gems/ONNX/Code/Source/Clients/Mnist.cpp

```diff
@@ -0,0 +1,176 @@
+    {
+        // Gets the png image from file and decodes using upng library.
+        upng_t* upng = upng_new_from_file(path);
+        upng_decode(upng);
+        const unsigned char* buffer = upng_get_buffer(upng);
+
+        // Converts image from buffer into binary greyscale representation.
+        // i.e. a pure black pixel is a 0, anything else is a 1.
+        // Bear in mind that the images in the dataset are flipped compared to how we'd usually think,
+        // so the background is black and the actual digit is white.
+        for (int y = 0; y < m_imageHeight; y++)
+        {
+            for (int x = 0; x < m_imageWidth; x++)
+            {
+                int content = static_cast<int>(buffer[(y)*m_imageWidth + x]);
+                if (content == 0)
+                {
+                    m_input[m_imageWidth * y + x] = 0.0f;
+                }
+                else
+                {
+                    m_input[m_imageHeight * y + x] = 1.0f;
+                }
+            }
+        }
+    }
+
+    MnistReturnValues MnistExample(MNIST& mnist, const char* path)
+    {
+        mnist.LoadImage(path);
+        mnist.Run(mnist.m_input, mnist.m_output);
+        mnist.GetResult();
+
+        MnistReturnValues returnValues;
+        returnValues.m_inference = mnist.m_result;
+        returnValues.m_runtime = mnist.m_delta;
+        return (returnValues);
+    }
+
+    void RunMnistSuite(int testsPerDigit, bool cudaEnable)
+    {
+        // Initialises and loads the mnist model.
+        // The same instance of the model is used for all runs.
+        MNIST mnist;
+        AZStd::vector<float> input(mnist.m_imageSize);
+        mnist.m_input = input;
+        AZStd::vector<float> output(10);
+        mnist.m_output = output;
+
+        MNIST::InitSettings modelInitSettings;
+        modelInitSettings.m_inputShape = { 1, 1, 28, 28 };
+        modelInitSettings.m_outputShape = { 1, 10 };
+
+        if (cudaEnable)
+        {
+            modelInitSettings.m_modelName = "MNIST CUDA (Precomputed)";
+            modelInitSettings.m_cudaEnable = true;
+        }
+        else
+        {
+            modelInitSettings.m_modelName = "MNIST (Precomputed)";
+        }
+
+        mnist.Load(modelInitSettings);
+
+        int numOfEach = testsPerDigit;
+        int totalFiles = 0;
+        int64_t numOfCorrectInferences = 0;
+        float totalRuntimeInMilliseconds = 0;
+
+        // This bit cycles through the folder with all the mnist test images, calling MnistExample() for the specified number of each digit.
+        // The structure of the folder is as such: /testing/{digit}/{random_integer}.png    e.g /testing/3/10.png
+        for (int digit = 0; digit < 10; digit++)
+        {
+            std::filesystem::directory_iterator iterator =
+                std::filesystem::directory_iterator{ std::wstring{ W_GEM_ASSETS_PATH } + L"/mnist_png/testing/" + std::to_wstring(digit) +
+                                                     L"/" };
```

Comment:
std::filesystem shouldn't be used within the O3DE codebase.

Furthermore the `W_GEM_ASSETS_PATH` and `GEM_ASSETS_PATH` defines should be removed.

What should be done here is to use the `AZ::IO::FileIOBase::ResolvePath` function to determine the actual file path to the ONNX gem root and then use `AZ::IO::SystemFile::FindFiles` locate all the files within a directory.

```c++
auto TestImage = [&mnist, testPerDigit, &totalFiles, &totalRuntimeInMilliseconds](AZ::IO::Path digitFilePath, bool isFile) -> bool
{
    if (!isFile)
    {
        AZ::IO::FixedMaxPath directoryName = digitFilePath.Filename();
        char* directoryEnd;
        AZ::s64 digit = strtoll(directoryName.c_str(), &directoryEnd, 10);
        if (directoryName.c_str() != directoryEnd)
        {
            // The current folder is the @gemroot:ONNX@/testing/{digit} folder
            // Search for any png files
            auto RunMnistExample = [&mnist, testPerDigit, &totalFiles, &totalRuntimeInMilliseconds, digit](AZ::IO::Path pngFilePath, bool) -> bool
            {
                MnistReturnValues returnedValues = MnistExample(mnist, pngFilePath.c_str());
                if (returnedValues.m_inference == digit)
                {
                    numOfCorrectInferences += 1;
                }
                mnist.DispatchTimingSample();
                totalRuntimeInMilliseconds += returnedValues.m_runtime;
                totalFiles++;
            };
            AZ::IO::SystemFile::FindFiles((digitFilePath / "*.png").c_str(), RunMnistExample);
        }
    }
};

// Get the FileIOBase to resolve the path to the ONNX gem
AZ::IO::FileIOBase* fileIo = AZ::IO::FileIOBase::GetInstance();
AZ::IO::FixedMaxPath onnxGemRoot;
if (fileIo->ResolvePath(onnxGemRoot, "gemroot:ONNX@))
{
    // onnxGemRoot is set to the root folder of the ONNX gem
   AZ::IO::SystemFile::FindFiles((onnxGemRoot / "Assets/testing/*").c_str(), TestImage);
}
```

Reply:
Thank you for this, I'm trying to change it at the moment. Let's say I didn't want to iterate over every single image in each digit folder, what would be the correct way of specifying only x number of images of each digit to be tested?

Reply:
I'm currently checking how many files in a digit folder have been run through an if statement, and then just not running MnistExample if exceeded. Is there a more efficient way of doing this without going into the callback for every file found?

```// This bit cycles through the folder with all the mnist test images, calling MnistExample() for the specified number of each digit.
        // The structure of the folder is as such: /testing/{digit}/{random_integer}.png    e.g /testing/3/10.png
        auto TestImage = [&mnist, &numOfCorrectInferences, &totalFiles, &totalRuntimeInMilliseconds, &mnistTestImageRoot, numOfEach](AZ::IO::Path digitFilePath, bool isFile) -> bool
        {
            if (!isFile)
            {
                AZ::IO::FixedMaxPath directoryName = digitFilePath.Filename();
                char* directoryEnd;
                AZ::s64 digit = strtoll(directoryName.c_str(), &directoryEnd, 10);
                if (directoryName.c_str() != directoryEnd)
                {
                    // How many files of that digit have been tested
                    int version = 0;

                    // Search for any png files
                    auto RunMnistExample = [&mnist, &numOfCorrectInferences, &totalFiles, &totalRuntimeInMilliseconds, &mnistTestImageRoot, &directoryName, &version, &digit, &numOfEach](AZ::IO::Path pngFilePath, bool) -> bool
                    {
                        // Stop running examples once version limit for that digit has been reached
                        if ((version < numOfEach))
                        {
                            MnistReturnValues returnedValues = MnistExample(mnist, (mnistTestImageRoot / directoryName / pngFilePath).c_str());
                            if (returnedValues.m_inference == digit)
                            {
                                numOfCorrectInferences += 1;
                            }
                            mnist.DispatchTimingSample();
                            totalRuntimeInMilliseconds += returnedValues.m_runtime;
                            totalFiles++;
                            version++;
                        }
                        return true;
                    };
                    AZ::IO::SystemFile::FindFiles((mnistTestImageRoot / directoryName / "*.png").c_str(), RunMnistExample);
                }
            }
            return true;
        };
```